Add new rule ST09: Joins should list the table referenced earlier (default)/later first

thibonacci opened the pull request:
<!--Thanks for adding this feature!-->

<!--Please give the Pull Request a meaningful title for the release notes-->

### Brief summary of the change made
<!--Please include `fixes #XXXX` to automatically close any corresponding issue when the pull request is merged. Alternatively if not fully closed you can say `makes progress on #XXXX`.-->

Fixes #2395.

### Are there any other side effects of this change that we should be aware of?

There shouldn't be any.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

greg-finley commented:
Thanks for the contribution! As a first step, could you look into the issues raised by the CI pipeline?

fmms commented:
@thibonacci very nice. I like the idea. Another thing I have bbeen wishing for is to require ON clauses to be in brackets.

Regarding the current implmentation, I think one has to consider how this behaves on `ÌNNER JOIN` and `RIGHT JOIN`. To my understanding this will even be a challenge for the naming as `right` and `left` get hard to understand? Maybe we have a `basetable` and a `jointable` table?

thibonacci commented:
> Thanks for the contribution! As a first step, could you look into the issues raised by the CI pipeline?

Thanks @greg-finley , I believe I've fixed all of the issues except for [this one](https://github.com/sqlfluff/sqlfluff/actions/runs/5573581086/job/15103595190) but it seems to only be affecting Python 3.7, which has reached end of life, so perhaps it's okay? If you could help rerun the CI pipeline and we can check if there are still any errors? Thanks!

thibonacci commented:
> @thibonacci very nice. I like the idea. Another thing I have bbeen wishing for is to require ON clauses to be in brackets.
> 
> Regarding the current implmentation, I think one has to consider how this behaves on `ÌNNER JOIN` and `RIGHT JOIN`. To my understanding this will even be a challenge for the naming as `right` and `left` get hard to understand? Maybe we have a `basetable` and a `jointable` table?

Yes that's not a bad point, I'm just a bit disappointed not be to able to find official documentation using terminology that we are looking for here. Will continue to look...

greg-finley commented:
> > Thanks for the contribution! As a first step, could you look into the issues raised by the CI pipeline?
> 
> Thanks @greg-finley , I believe I've fixed all of the issues except for [this one](https://github.com/sqlfluff/sqlfluff/actions/runs/5573581086/job/15103595190) but it seems to only be affecting Python 3.7, which has reached end of life, so perhaps it's okay? If you could help rerun the CI pipeline and we can check if there are still any errors? Thanks!

For consistency with the rest of the codebase, let's do `from typing import List` and `List[str]` etc.

thibonacci commented:
> > > Thanks for the contribution! As a first step, could you look into the issues raised by the CI pipeline?
> > 
> > 
> > Thanks @greg-finley , I believe I've fixed all of the issues except for [this one](https://github.com/sqlfluff/sqlfluff/actions/runs/5573581086/job/15103595190) but it seems to only be affecting Python 3.7, which has reached end of life, so perhaps it's okay? If you could help rerun the CI pipeline and we can check if there are still any errors? Thanks!
> 
> For consistency with the rest of the codebase, let's do `from typing import List` and `List[str]` etc.

Thanks @greg-finley , I've done that and fixed more CI/CD errors. Hopefully only the py3.7 one is left.

thibonacci commented:
> [![Coverage Status](https://camo.githubusercontent.com/6bae0a4926054e11766bea69894190bdf8f71dc9f354efafaf8d19f2959491af/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f36313638323532362f6261646765)](https://coveralls.io/builds/61682526)
> 
> coverage: 99.952% (-0.05%) from 100.0% when pulling **[83f40c7](https://github.com/sqlfluff/sqlfluff/commit/83f40c7558293e282a220f62daddc4be07d980e7) on thibonacci:main** into **[a4563fa](https://github.com/sqlfluff/sqlfluff/commit/a4563fa30b2da435fc1541b717fa9648d4c91b97) on sqlfluff:main**.

@greg-finley I'm working on getting the coverage fixed; however, I'm now getting pre-commit errors about files that I've never worked on (see below screenshot). Do you have any idea as to what I should do in order to solve them? Thanks!
![Screenshot 2023-08-01 at 08 23 33](https://github.com/sqlfluff/sqlfluff/assets/60314995/6a8a8289-a32d-4593-aafb-8cfbaa065bb8)

alanmcruickshank commented:
Thanks for getting this so far @thibonacci!

Having a read through I think this is a good rule for us to add - but we need to cover a couple of logical situations to make sure it's complete and ready to add.

1. I think the point raised by @fmms is important, and we should at least one test case each for the following joins:
   - `LEFT` (which we have)
   - `LEFT OUTER` - which should behave the same as `LEFT`
   - `INNER` - which should also behave the same as `LEFT`
   - `RIGHT` & `RIGHT OUTER` - which _may_ behave differently. I've done a bit of research, and based on the [snowflake docs](https://docs.snowflake.com/en/sql-reference/constructs/join) and the [bigquery docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#right_join) it appears there is disagreement on which should come first - but that it's the _same_ for both `RIGHT` and `LEFT`. That means I think that from a configuration point of view, while we might still offer `left` and `right` - the default should be something more like `earlier` and `later` (I'm not attached to those terms, maybe `source` and `join`?), and the default should be `earlier`/`source` in line with #2395.
   - We should also test with `CROSS` and `FULL OUTER`.
2. We should also address the question of joins which do their filtering in the `WHERE` clause. My sense is that you haven't tried to tackle this as this stage, which is fine, but we should put a note in the docstring to that effect to say it's not covered. e.g. `SELECT * FROM a, b WHERE a.foo = b.foo`.
  - On this front, it could be interesting to also extend this logic to the `WHERE` clause too though, where we say any comparisons in there should _also_ follow the same rules. Up to you whether you apply it to `WHERE` for now, or just call out that it's not currently covered.

alanmcruickshank commented:
Also heads up that we'll probably get 2.2.0 release before this is ready, don't rush this to try and hit that timeline. We can get this into 2.3.0 and I'm happy to do that as soon as this is ready to go 👍 .

thibonacci commented:
> Thanks for getting this so far @thibonacci!
> 
> Having a read through I think this is a good rule for us to add - but we need to cover a couple of logical situations to make sure it's complete and ready to add.
> 
> 1. I think the point raised by @fmms is important, and we should at least one test case each for the following joins:
>    
>    * `LEFT` (which we have)
>    * `LEFT OUTER` - which should behave the same as `LEFT`
>    * `INNER` - which should also behave the same as `LEFT`
>    * `RIGHT` & `RIGHT OUTER` - which _may_ behave differently. I've done a bit of research, and based on the [snowflake docs](https://docs.snowflake.com/en/sql-reference/constructs/join) and the [bigquery docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#right_join) it appears there is disagreement on which should come first - but that it's the _same_ for both `RIGHT` and `LEFT`. That means I think that from a configuration point of view, while we might still offer `left` and `right` - the default should be something more like `earlier` and `later` (I'm not attached to those terms, maybe `source` and `join`?), and the default should be `earlier`/`source` in line with [New Rule: In ON clause of JOIN statement specify source table first and join table second #2395](https://github.com/sqlfluff/sqlfluff/issues/2395).
>    * We should also test with `CROSS` and `FULL OUTER`.
> 2. We should also address the question of joins which do their filtering in the `WHERE` clause. My sense is that you haven't tried to tackle this as this stage, which is fine, but we should put a note in the docstring to that effect to say it's not covered. e.g. `SELECT * FROM a, b WHERE a.foo = b.foo`.
> 
> * On this front, it could be interesting to also extend this logic to the `WHERE` clause too though, where we say any comparisons in there should _also_ follow the same rules. Up to you whether you apply it to `WHERE` for now, or just call out that it's not currently covered.

Thanks for taking a look @alanmcruickshank!

1. Yep will add those tests! Still not quite fully sure about the terminology, Redshift actually does use "left" and "right" in [here](https://docs.aws.amazon.com/redshift/latest/dg/r_FROM_clause30.html#r_FROM_clause30-parameters) the way I've been using them (i.e. irrespective of the type of join) - in quotes though so they're essentially admitting that it's not the greatest terminology. I'm happy to go with `source` and `join`, just wish there were some authority telling us what we should use, but I suppose SQLFluff is becoming such an authority!
2. Yeah when I first started this I had no idea how SQLFluff worked so wanted to keep things as simple as possible but I now don't think it should be too hard to extend the logic I've used to the `WHERE` clause so I can do that as part of this PR.

I have very limited time to work on this unfortunately so am definitely not rushing anything 👍

thibonacci commented:
@alanmcruickshank In the end I decided not to try to cover `WHERE` clauses as I've just learned about the `(+)` notation (see [Snowflake docs](https://docs.snowflake.com/en/sql-reference/constructs/where#joins-in-the-where-clause)) which I don't believe is supported by SQLFluff yet. Also went with `earlier`/`later` as it felt a little bit easier to explain in the docstring. 
If you could help trigger the CI which I think should be all green now. 🤞 
Thanks!

alanmcruickshank commented on src/sqlfluff/rules/structure/ST09.py:
```diff
@@ -0,0 +1,284 @@
+
+        # STEP 5.b.
+        else:
+            return LintResult(anchor=context.segment, fixes=fixes)
```
One last suggestion here. The current description shown in the test suite is `Joins should list the table referenced earlier/later first`. By setting the `description` argument here you can reflect the current config:

```suggestion
            return LintResult(
                anchor=context.segment,
                fixes=fixes,
                description=f"Joins should list the table referenced {self.preferred_first_table_in_join_clause} first"
            )
```

alanmcruickshank reviewed (commented):
Looking pretty good! One last suggestion on customising the `description` of the rule now that you've got all the configuration sorted (and you'll likely need to update the test suite too once you've done that).

After that, this should be good to merge.

thibonacci commented:
> Looking pretty good! One last suggestion on customising the `description` of the rule now that you've got all the configuration sorted (and you'll likely need to update the test suite too once you've done that).
> 
> After that, this should be good to merge.

Yep that's a nice little improvement. Done!

alanmcruickshank commented on src/sqlfluff/rules/structure/ST09.py:
```diff
@@ -0,0 +1,291 @@
+    """
+
+    name = "structure.first_table"
+    aliases = ("L070",)
```
No need to add an alias, that's only for the older rules which did have old names. The plan isn't to continue supporting that format going forward.

```suggestion
    aliases = ()
```

alanmcruickshank reviewed (changes requested):
Sorry, found one more thing on my last run through before merging. Hopefully an easy one. After that we're good to merge. I'd love to get this into the next release.

alanmcruickshank reviewed (approved):
Epic. Let's go! 🚀 

Thanks for taking the time to put this PR together. This will be in the next release.

fmms commented:
I have just tried this on several hunded files. Awesome.

One exception though:
```
PS C:\Users\fmoeller\Desktop\RepositoryVorlage> C:\Users\fmoeller\AppData\Roaming\Python\Python311\Scripts\sqlfluff.exe fix --config .\.sqlfluff --rules ST09 ".\50 SSDT\SSDT_DWH\UTIL_SSIS\Stored Procedures\get_events_with_duration.sql"
==== finding fixable violations ====
CRITICAL   [ST09] Applying rule ST09 to '50 SSDT\\SSDT_DWH\\UTIL_SSIS\\Stored Procedures\\get_events_with_duration.sql' threw an Exception: 'NoneType' object has no attribute 'raw_upper'
Traceback (most recent call last):
  File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\core\rules\base.py", line 866, in crawl
    res = self._eval(context=context)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\rules\structure\ST09.py", line 181, in _eval
    first_table = first_column_reference.get_child("naked_identifier").raw_upper
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'raw_upper'
==== no fixable linting violations found ====
All Finished 📜 🎉!
  [1 unfixable linting violations found]
```

unfortunately a huge file that I cannot share publically. Any idea how to get the line of the SQL so that i can create an example for you?  Otherwise I can mail it to one of you.

alanmcruickshank commented:
if I were to guess, it looks like it's failing to find any identifiers on
the join clause (and so get_child returns None).

We should make sure this is robust to statements where the join clause is
only literals for example, e.g:

`select * from a join b on (1 = 2)`

This might also be one where there _is_ a child but not directly:

`select * from a join b on (date_trunc('day', a.a) = date_trunc('day',
b.b))`

If you're able to make a new PR with this test cases added, I'm fairly
hopeful you'll be able to recreate the issue.

On Fri, 11 Aug 2023, 08:57 thibonacci, ***@***.***> wrote:

> I have just tried this on several hunded files. Awesome.
>
> One exception though:
>
> PS C:\Users\fmoeller\Desktop\RepositoryVorlage> C:\Users\fmoeller\AppData\Roaming\Python\Python311\Scripts\sqlfluff.exe fix --config .\.sqlfluff --rules ST09 ".\50 SSDT\SSDT_DWH\UTIL_SSIS\Stored Procedures\get_events_with_duration.sql"
> ==== finding fixable violations ====
> CRITICAL   [ST09] Applying rule ST09 to '50 SSDT\\SSDT_DWH\\UTIL_SSIS\\Stored Procedures\\get_events_with_duration.sql' threw an Exception: 'NoneType' object has no attribute 'raw_upper'
> Traceback (most recent call last):
>   File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\core\rules\base.py", line 866, in crawl
>     res = self._eval(context=context)
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\rules\structure\ST09.py", line 181, in _eval
>     first_table = first_column_reference.get_child("naked_identifier").raw_upper
>                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> AttributeError: 'NoneType' object has no attribute 'raw_upper'
> ==== no fixable linting violations found ====
> All Finished 📜 🎉!
>   [1 unfixable linting violations found]
>
> unfortunately a huge file that I cannot share publically. Any idea how to
> get the line of the SQL so that i can create an example for you? Otherwise
> I can mail it to one of you.
>
> @fmms <https://github.com/fmms> Thanks for this. If it's several SQL
> statements I'd just remove some until I find which one it is. If it's one
> single statement I'd remove join subconditions until I find which one it
> is. It's also most likely a pretty weird edge case. Happy to take a look as
> well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sqlfluff/sqlfluff/pull/4974#issuecomment-1674349032>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDULOHWZ672DCHQPL5KEXDXUXQVBANCNFSM6AAAAAA2MTTJ4Q>
> .
> You are receiving this because you modified the open/close state.Message
> ID: ***@***.***>
>

thibonacci commented:
> if I were to guess, it looks like it's failing to find any identifiers on the join clause (and so get_child returns None). We should make sure this is robust to statements where the join clause is only literals for example, e.g: `select * from a join b on (1 = 2)` This might also be one where there _is_ a child but not directly: `select * from a join b on (date_trunc('day', a.a) = date_trunc('day', b.b))` If you're able to make a new PR with this test cases added, I'm fairly hopeful you'll be able to recreate the issue.
> […](#)
> On Fri, 11 Aug 2023, 08:57 thibonacci, ***@***.***> wrote: I have just tried this on several hunded files. Awesome. One exception though: PS C:\Users\fmoeller\Desktop\RepositoryVorlage> C:\Users\fmoeller\AppData\Roaming\Python\Python311\Scripts\sqlfluff.exe fix --config .\.sqlfluff --rules ST09 ".\50 SSDT\SSDT_DWH\UTIL_SSIS\Stored Procedures\get_events_with_duration.sql" ==== finding fixable violations ==== CRITICAL [ST09] Applying rule ST09 to '50 SSDT\\SSDT_DWH\\UTIL_SSIS\\Stored Procedures\\get_events_with_duration.sql' threw an Exception: 'NoneType' object has no attribute 'raw_upper' Traceback (most recent call last): File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\core\rules\base.py", line 866, in crawl res = self._eval(context=context) ^^^^^^^^^^^^^^^^^^^^^^^^^^^ File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\rules\structure\ST09.py", line 181, in _eval first_table = first_column_reference.get_child("naked_identifier").raw_upper ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ AttributeError: 'NoneType' object has no attribute 'raw_upper' ==== no fixable linting violations found ==== All Finished 📜 🎉! [1 unfixable linting violations found] unfortunately a huge file that I cannot share publically. Any idea how to get the line of the SQL so that i can create an example for you? Otherwise I can mail it to one of you. @fmms <https://github.com/fmms> Thanks for this. If it's several SQL statements I'd just remove some until I find which one it is. If it's one single statement I'd remove join subconditions until I find which one it is. It's also most likely a pretty weird edge case. Happy to take a look as well. — Reply to this email directly, view it on GitHub <[#4974 (comment)](https://github.com/sqlfluff/sqlfluff/pull/4974#issuecomment-1674349032)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABDULOHWZ672DCHQPL5KEXDXUXQVBANCNFSM6AAAAAA2MTTJ4Q> . You are receiving this because you modified the open/close state.Message ID: ***@***.***>

Thanks @alanmcruickshank , it doesn't seem like any of those two tests are causing a critical error. They both get ignored by ST09 as there aren't any qualified column reference, comparison operator and qualified column reference sequences (although I have thought about extending the rule to cases where the column reference is wrapped inside a function).

However, I can see that a critical error does appear when both the table alias and the column name are quoted, e.g.:
`select "foo"."a" from "foo" join "bar" on ("bar"."b" = "foo"."b");`
in which case the column references indeed do not have naked_identifier children.

Will try to raise a PR something asap 🙏

thibonacci commented:
> @thibonacci i have gone through a thousand lines of SQL and finally found a minimal example:
> 
> ```sql
>   SELECT * FROM [tab1] t LEFT JOIN tab2 ON ([t].[a] = tab2.[a])
> ```

Thanks @fmms, what dialect is that?

fmms commented:
@thibonacci tsql ...

thibonacci commented:
> @thibonacci tsql ...

Thanks @fmms, sorry never used tsql before. So I think that #5087 should ensure we're not getting any more critical errors but your example will be ignored by ST09 as technically `[t]` is not in the list of table aliases that were found - which is `['t', 'tab2']` but the following will be fixed - assuming you're using the `later` configuration:
```
SELECT * FROM [tab1] t LEFT JOIN tab2 ON (t.[a] = tab2.[a])
```
I'll try to think whether ST09 can support your exact example.

fmms commented:
Tested the new code:

```
WITH taba AS (
  SELECT 1 AS c
)

, tabb AS (
  SELECT 1 AS c
)

SELECT * FROM taba
LEFT JOIN tabb ON (taba.c = [tabb].[c])
```

That is a totally valid t-SQL Statement.

However, it fails:
```
PS C:\Users\fmoeller\Desktop\RepositoryVorlage> C:\Users\fmoeller\AppData\Roaming\Python\Python311\Scripts\sqlfluff.exe fix --config .\.sqlfluff .\query.sql
==== finding fixable violations ====
CRITICAL   [ST09] Applying rule ST09 to 'query.sql' threw an Exception: 'NoneType' object has no attribute 'raw_upper'
Traceback (most recent call last):
  File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\core\rules\base.py", line 866, in crawl
    res = self._eval(context=context)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\fmoeller\AppData\Roaming\Python\Python311\site-packages\sqlfluff\rules\structure\ST09.py", line 184, in _eval
    ).raw_upper
      ^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'raw_upper'
==== no fixable linting violations found ====
All Finished 📜 🎉!
  [3 unfixable linting violations found]
```

the parse looks like the following:
```
PS C:\Users\fmoeller\Desktop\RepositoryVorlage> C:\Users\fmoeller\AppData\Roaming\Python\Python311\Scripts\sqlfluff.exe parse --config .\.sqlfluff  .\query.sql
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    batch:
[L:  1, P:  1]      |        statement:
[L:  1, P:  1]      |            with_compound_statement:
[L:  1, P:  1]      |                keyword:                                      'WITH'
[L:  1, P:  5]      |                whitespace:                                   ' '
[L:  1, P:  6]      |                common_table_expression:
[L:  1, P:  6]      |                    naked_identifier:                         'taba'
[L:  1, P: 10]      |                    whitespace:                               ' '
[L:  1, P: 11]      |                    keyword:                                  'AS'
[L:  1, P: 13]      |                    whitespace:                               ' '
[L:  1, P: 14]      |                    bracketed:
[L:  1, P: 14]      |                        start_bracket:                        '('
[L:  1, P: 15]      |                        [META] indent:
[L:  1, P: 15]      |                        newline:                              '\n'
[L:  2, P:  1]      |                        whitespace:                           '  '
[L:  2, P:  3]      |                        select_statement:
[L:  2, P:  3]      |                            select_clause:
[L:  2, P:  3]      |                                keyword:                      'SELECT'
[L:  2, P:  9]      |                                [META] indent:
[L:  2, P:  9]      |                                whitespace:                   ' '
[L:  2, P: 10]      |                                select_clause_element:
[L:  2, P: 10]      |                                    numeric_literal:          '1'
[L:  2, P: 11]      |                                    whitespace:               ' '
[L:  2, P: 12]      |                                    alias_expression:
[L:  2, P: 12]      |                                        [META] indent:
[L:  2, P: 12]      |                                        keyword:              'AS'
[L:  2, P: 14]      |                                        whitespace:           ' '
[L:  2, P: 15]      |                                        naked_identifier:     'c'
[L:  2, P: 16]      |                                        [META] dedent:
[L:  2, P: 16]      |                            [META] dedent:
[L:  2, P: 16]      |                        newline:                              '\n'
[L:  3, P:  1]      |                        [META] dedent:
[L:  3, P:  1]      |                        end_bracket:                          ')'
[L:  3, P:  2]      |                newline:                                      '\n'
[L:  4, P:  1]      |                newline:                                      '\n'
[L:  5, P:  1]      |                comma:                                        ','
[L:  5, P:  2]      |                whitespace:                                   ' '
[L:  5, P:  3]      |                common_table_expression:
[L:  5, P:  3]      |                    naked_identifier:                         'tabb'
[L:  5, P:  7]      |                    whitespace:                               ' '
[L:  5, P:  8]      |                    keyword:                                  'AS'
[L:  5, P: 10]      |                    whitespace:                               ' '
[L:  5, P: 11]      |                    bracketed:
[L:  5, P: 11]      |                        start_bracket:                        '('
[L:  5, P: 12]      |                        [META] indent:
[L:  5, P: 12]      |                        newline:                              '\n'
[L:  6, P:  1]      |                        whitespace:                           '  '
[L:  6, P:  3]      |                        select_statement:
[L:  6, P:  3]      |                            select_clause:
[L:  6, P:  3]      |                                keyword:                      'SELECT'
[L:  6, P:  9]      |                                [META] indent:
[L:  6, P:  9]      |                                whitespace:                   ' '
[L:  6, P: 10]      |                                select_clause_element:
[L:  6, P: 10]      |                                    numeric_literal:          '1'
[L:  6, P: 11]      |                                    whitespace:               ' '
[L:  6, P: 12]      |                                    alias_expression:
[L:  6, P: 12]      |                                        [META] indent:
[L:  6, P: 12]      |                                        keyword:              'AS'
[L:  6, P: 14]      |                                        whitespace:           ' '
[L:  6, P: 15]      |                                        naked_identifier:     'c'
[L:  6, P: 16]      |                                        [META] dedent:
[L:  6, P: 16]      |                            [META] dedent:
[L:  6, P: 16]      |                        newline:                              '\n'
[L:  7, P:  1]      |                        [META] dedent:
[L:  7, P:  1]      |                        end_bracket:                          ')'
[L:  7, P:  2]      |                newline:                                      '\n'
[L:  8, P:  1]      |                newline:                                      '\n'
[L:  9, P:  1]      |                select_statement:
[L:  9, P:  1]      |                    select_clause:
[L:  9, P:  1]      |                        keyword:                              'SELECT'
[L:  9, P:  7]      |                        [META] indent:
[L:  9, P:  7]      |                        whitespace:                           ' '
[L:  9, P:  8]      |                        select_clause_element:
[L:  9, P:  8]      |                            wildcard_expression:
[L:  9, P:  8]      |                                wildcard_identifier:
[L:  9, P:  8]      |                                    star:                     '*'
[L:  9, P:  9]      |                        whitespace:                           ' '
[L:  9, P: 10]      |                    [META] dedent:
[L:  9, P: 10]      |                    from_clause:
[L:  9, P: 10]      |                        keyword:                              'FROM'
[L:  9, P: 14]      |                        whitespace:                           ' '
[L:  9, P: 15]      |                        from_expression:
[L:  9, P: 15]      |                            [META] indent:
[L:  9, P: 15]      |                            from_expression_element:
[L:  9, P: 15]      |                                table_expression:
[L:  9, P: 15]      |                                    table_reference:
[L:  9, P: 15]      |                                        naked_identifier:     'taba'
[L:  9, P: 19]      |                            newline:                          '\n'
[L: 10, P:  1]      |                            [META] dedent:
[L: 10, P:  1]      |                            join_clause:
[L: 10, P:  1]      |                                keyword:                      'LEFT'
[L: 10, P:  5]      |                                whitespace:                   ' '
[L: 10, P:  6]      |                                keyword:                      'JOIN'
[L: 10, P: 10]      |                                [META] indent:
[L: 10, P: 10]      |                                whitespace:                   ' '
[L: 10, P: 11]      |                                from_expression_element:
[L: 10, P: 11]      |                                    table_expression:
[L: 10, P: 11]      |                                        table_reference:
[L: 10, P: 11]      |                                            naked_identifier:  'tabb'
[L: 10, P: 15]      |                                whitespace:                   ' '
[L: 10, P: 16]      |                                [META] dedent:
[L: 10, P: 16]      |                                [META] indent:
[L: 10, P: 16]      |                                join_on_condition:
[L: 10, P: 16]      |                                    keyword:                  'ON'
[L: 10, P: 18]      |                                    [META] (implicit) indent:
[L: 10, P: 18]      |                                    whitespace:               ' '
[L: 10, P: 19]      |                                    bracketed:
[L: 10, P: 19]      |                                        start_bracket:        '('
[L: 10, P: 20]      |                                        [META] indent:
[L: 10, P: 20]      |                                        expression:
[L: 10, P: 20]      |                                            column_reference:
[L: 10, P: 20]      |                                                naked_identifier:  'tabb'
[L: 10, P: 24]      |                                                dot:          '.'
[L: 10, P: 25]      |                                                naked_identifier:  'c'
[L: 10, P: 26]      |                                            whitespace:       ' '
[L: 10, P: 27]      |                                            comparison_operator:
[L: 10, P: 27]      |                                                raw_comparison_operator:  '='
[L: 10, P: 28]      |                                            whitespace:       ' '
[L: 10, P: 29]      |                                            column_reference:
[L: 10, P: 29]      |                                                quoted_identifier:  '[taba]'
[L: 10, P: 35]      |                                                dot:          '.'
[L: 10, P: 36]      |                                                quoted_identifier:  '[c]'
[L: 10, P: 39]      |                                        [META] dedent:
[L: 10, P: 39]      |                                        end_bracket:          ')'
[L: 10, P: 40]      |                                    [META] dedent:
[L: 10, P: 40]      |                                [META] dedent:
[L: 10, P: 40]      |    newline:                                                  '\n'
[L: 11, P:  1]      |    [META] end_of_file:
```

thibonacci commented:
@fmms Thanks. I believe you shouldn't be getting any critical errors any longer as long as you're using the fix from #5087. But as I said, for your example ST09 will just ignore your join condition, which I guess isn't what you want. 
Here's the situation as I see it:
- essentially ST09 currently doesn't support cases where there is inconsistent quoting, e.g. using `taba` first and then `[taba]` in the join condition
- if you turn RF06 on then I think it actually becomes supported since RF06 makes quoting consistent and seems to be run before ST09, but I'm guessing you and potentially other people are not using that rule
- I potentially have a "fix" to allow ST09 to support inconsistent quoting and while it seems to be working I'm not too comfortable with it as quotes are treated quite differently depending on the dialect
Therefore I think it might be good to open a new issue and then I can raise a PR referencing that issue.

fmms commented:
Are you really not getting the exception? I am on most recent git Version and still got the error. 

I am quiet confident to be on that version as you see that the exception shows the reformatted code. 

To my testing the exception happens if alias AND column name are in quotes []…

thibonacci commented:
@fmms I _was_ getting it but am not getting it now 🤔 . Looking at the parsed SQL, you have two different column_reference segments and they both have at least one child who is either a `naked_identifier` segment or a `quoted_identifier` segment, and we should be able to use `raw` on those segments.
